### PR TITLE
Causal LM generation fix for prefix tuning: GPT2 model

### PR DESCRIPTION
**Pull request by vineetm**
Hi @pacman100 

Causal LM generation with prefix tuning fails for GPT2 models. This is due to mismatch in size in input_ids and position_ids (which GPT2 model creates on the fly). This, PR fixes it. Please have a look.

Minimal code to reproduce the issue:

```python
import torch
from transformers import AutoModelForCausalLM
from peft import PrefixTuningConfig, TaskType, get_peft_model


model_name_or_path = "gpt2"
peft_config = PrefixTuningConfig(task_type=TaskType.CAUSAL_LM, num_virtual_tokens=3)
model = AutoModelForCausalLM.from_pretrained(model_name_or_path)
model = get_peft_model(model, peft_config)

inputs = {
    'input_ids': torch.tensor([1, 2]).unsqueeze(0),
    'attention_mask': torch.tensor([1, 1]).unsqueeze(0),
}

outputs = model.generate(
    **inputs, max_new_tokens=20, eos_token_id=3
)
print(outputs)
```

### CODE REVIEW

**Review by younesbelkada — APPROVED** [review]
Awesome contribution! 
Thanks a mile for fixing, I can confirm with the suggested change, all `generate` tests pass in #223

File: src/peft/peft_model.py [hunk 1]
```diff
@@ -582,7 +582,13 @@ def generate(self, **kwargs):
             else:
                 if "input_ids" not in kwargs:
                     raise ValueError("input_ids must be provided for Peft model generation")
-                if kwargs.get("attention_mask", None) is not None:
+                # For gpt2 models, we construct postion_ids on the fly by using attention mask, and position ids need to match input_shape.
+                # for prefix tuning, input shape is determined using `input_ids`. Thus we should not expand 'attention_mask' here
+                # for prompt tuning input_ids is not passed but a concatenated input_embeds is passed. Thus attention_mask needs to be of same size of num_virtual_tokens + input_ids
+                if (
+                    kwargs.get("attention_mask", None) is not None
+                    and self.peft_config.peft_type == PeftType.PROMPT_TUNING
```
[the remark this answers is not in the document]
Comment (younesbelkada) [reply src/peft/peft_model.py:590]: otherwise `generate` will fail for `P_TUNING` models

**Comment by vineetm** [comment]
Thank you @younesbelkada! I have made the suggested change.

**Comment by younesbelkada** [comment]
Don't worry about the failing styling check , it will be fixed soon #224 , we'll merge this right after

**Review by pacman100 — APPROVED** [review]
Thank you @vineetm for fixing the pesky bug, LGTM!  🤗

**Comment by vineetm** [comment]
@younesbelkada I have merged my branch with main. Thanks!

**Comment by younesbelkada** [comment]
Awesome, thanks again for your great work on this!

**Comment by vineetm** [comment]
@younesbelkada @pacman100 You are most welcome! Thank you for creating this great repo!

**Pull request merged**